### PR TITLE
LPS-149266 test changes

### PR DESCRIPTION
**Pull request by dantewang**
@jeffwu0724 I made several SF changes to the test, however, I didn't finish my modification since I have some questions about your original test logic. I'll comment in this PR.

### CODE REVIEW

File: portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java [hunk 1]
```diff
@@ -223,6 +236,134 @@ public void testStripSuffixWhenFileNameHasParenthesisAtStart() {
...
+			_canFileReadWrite(_getTempTestFilePath("zip/test/directory")));
+
+		Assert.assertTrue(
+			_canFileReadWrite(
+				_getTempTestFilePath("zip/test/entry/entry.txt")));
+		_tearDown();
+	}
+
+	@Test
+	public void testUnzipZipSlipVulnerable() throws Exception {
```
Comment (dantewang) [inline portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java:266]: @jeffwu0724 For the vulnerable test, I think you should add assertion for the warning log output. On one hand this ensures the log is printed as expected, on the other hand it can avoid the warning to be printed to the console.

For example, when i run the unit test locally I see the following output:

```
    [junit] Running com.liferay.portal.util.FileImplTest
    [junit] Testsuite: com.liferay.portal.util.FileImplTest
    [junit] SLF4J: Class path contains multiple SLF4J bindings.
    [junit] SLF4J: Found binding in [file:/home/repo/liferay/portal-ce/util-slf4j/classes/org/slf4j/impl/StaticLoggerBinder.class]
    [junit] SLF4J: Found binding in [jar:file:/home/repo/liferay/portal-ce/lib/development/slf4j-simple.jar!/org/slf4j/impl/StaticLoggerBinder.class]
    [junit] SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
    [junit] Loading file:/home/repo/liferay/portal-ce/portal-impl/classes/system.properties
    [junit] SLF4J: Actual binding is of type [com.liferay.util.slf4j.LiferayLoggerFactory]
    [junit] Loading file:/home/repo/liferay/portal-ce/portal-impl/classes/portal.properties
    [junit] Loading file:/home/repo/liferay/portal-ce/portal-impl/test-classes/unit/portal-test.properties
    [junit] Tests run: 28, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.264 sec
    [junit] Tests run: 28, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.264 sec
    [junit] 
    [junit] 六月 07, 2022 5:42:09 上午 com.liferay.portal.kernel.log.Jdk14LogImpl warn
    [junit] 警告: Entry is outside of the target dir: ../../../../../../bad.txt
[stopwatch] [test.cmd: 0.807 sec]
```

The "警告" line is the logger's warning output, localized by my OS configuration.

You can take https://github.com/liferay-core-infra/liferay-portal/pull/873/commits/0f562908bddb2da7230e3664c9e919d5d83ed724#diff-5f9a1cead69535423c815c23f96fc45737baee11a8a5604ae2c7d49f48e9436fR253 as a reference.

File: portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java [hunk 2]
```diff
@@ -223,6 +236,134 @@ public void testStripSuffixWhenFileNameHasParenthesisAtStart() {
...
+			HashMapBuilder.put(
+				"../../../../../../bad.txt", "I am bad!"
+			).put(
+				"good.txt", "I am good!"
+			).build());
+
+		_fileImpl.unzip(file, _tempFolder);
+
+		_assertExists("good.txt");
+		_assertDoesNotExist("tmp/bad.txt");
```
Comment (dantewang) [inline portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java:280]: @jeffwu0724 I have a question about this assertion. The bad file has a path like `../../../../../../bad.txt`, but why is the assertion asserting `$_tempFolder/tmp/bad.txt`? It seems the zip content is just unzipped to the `_tempFolder`, so the bad file should be out of it, if the fix is not applied?

Could you double check this part, and make sure it fails without the `FileImpl` fix, while passes with the fix?

File: portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java [hunk 3]
```diff
@@ -223,6 +236,134 @@ public void testStripSuffixWhenFileNameHasParenthesisAtStart() {
...
+		Assert.assertFalse(Files.exists(fullPath));
+	}
+
+	private void _assertExists(String name) throws Exception {
+		Path fullPath = _getTempTestFilePath(name);
+
+		Assert.assertTrue(Files.exists(fullPath));
+	}
+
+	private boolean _canFileReadWrite(Path path) {
```
Comment (dantewang) [inline portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java:296]: @jeffwu0724 Why is this needed and asserted? By the way, why does it contain different logic for Windows and other OS?

File: portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java [hunk 4]
```diff
@@ -223,6 +236,134 @@ public void testStripSuffixWhenFileNameHasParenthesisAtStart() {
...
+		_assertExists("zip/test/directory");
+		_assertExists("zip/test/entry/entry.txt");
+
+		Assert.assertTrue(
+			_canFileReadWrite(_getTempTestFilePath("zip/test/directory")));
+
+		Assert.assertTrue(
+			_canFileReadWrite(
+				_getTempTestFilePath("zip/test/entry/entry.txt")));
+		_tearDown();
```
Comment (dantewang) [inline portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java:262]: @jeffwu0724 When any of the operation or assertion fails, the tear down method won't be executed. It should at least be protected using `try-finally` or `file.deleteOnExit()`.

File: portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java (hunk 2 above)
Comment (jeffwu0724) [reply portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java:280]: Hey Dante, seems like if I don't have the fix, it will show 
`Testcase: testUnzipZipSlipVulnerable(com.liferay.portal.util.FileImplTest): Caused an ERROR
    [junit] /tmp/20220607215702662XSIATEBR/../../../../../../bad.txt
    [junit] java.nio.file.AccessDeniedException: /tmp/20220607215702662XSIATEBR/../../../../../../bad.txt
    [junit]     at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84)
    [junit]     at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
    [junit]     at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
    [junit]     at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
    [junit]     at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
    [junit]     at java.nio.file.Files.newOutputStream(Files.java:216)
    [junit]     at java.nio.file.Files.copy(Files.java:3016)
    [junit]     at com.liferay.portal.util.FileImpl.unzip(FileImpl.java:849)
    [junit]     at com.liferay.portal.util.FileImplTest.testUnzipZipSlipVulnerable(FileImplTest.java:263)
    [junit]     at com.liferay.portal.kernel.test.rule.AbstractTestRule$2.evaluate(AbstractTestRule.java:99)
    [junit]     at com.liferay.portal.kernel.test.rule.AbstractTestRule$1.evaluate(AbstractTestRule.java:`
instead of just go to the _assertDoesNotExist("tmp/bad.txt");. I have changed my logic here according to the reference you sent me.

File: portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java (hunk 3 above)
Comment (jeffwu0724) [reply portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java:296]: For the OS checking, in the beginning I am wondering there will be different path format for different OS. I will remove it now, and I will check it with `ci:test:environment` and see if we need this part or not.

File: portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java (hunk 2 above)
Comment (dantewang) [reply portal-impl/test/unit/com/liferay/portal/util/FileImplTest.java:280]: @jeffwu0724 Yeah, here the error is because the Java process tries to extract this file to `/`, which is not writable unless you `sudo`. If you look at the fix and test made by Search Team, you'll see that they created the bad file to be `../../../../../../../../../../../../../tmp/bad.txt`. I guess this ensures the file to be extracted into `/tmp`.

However, `/tmp` only exists in *Unix. If you run it on Windows, I'm not sure about the behavior. I think the best way to test it is to create the file to be `../bad.txt` and extract it into a sub-dir under the temp dir we use. So without the fix, the file will be extracted into the temp dir and the test can easily find it.
